### PR TITLE
Add a helper to generate xy coordinates for AxisArtistHelper.

### PR DESCRIPTION
**Pull request by anntzer**
AxisArtistHelper can generate either x or y ticks/gridlines depending
on the value of self.nth_coord.  The implementation often requires
generating e.g. shape (2,) arrays such that the nth_coord column is
set to a tick position, and the 1-nth_coord column has is set to
0.  This is currently done using constructs like ``verts = [0, 0];
verts[self.nth_coord] = value`` where the mutation doesn't really help
legibility.  Instead, introduce a ``_to_xy`` helper that allows writing
``to_xy(variable=x, fixed=0)``.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/mpl_toolkits/axisartist/axislines.py [hunk 1]
```diff
@@ -108,41 +108,43 @@ def update_lim(self, axes):
         delta2 = _api.deprecated("3.6")(
             property(lambda self: 0.00001, lambda self, value: None))
 
+        def _to_xy(self, *, variable, fixed):
```
Comment (timhoffm) [inline lib/mpl_toolkits/axisartist/axislines.py:111]: It's good to make that concept explicit instead of wildly juggling indices in the code. Still, I find this quite hard to follow. Can we move this up one level in the semantics:

```
def _to_xy(self, values, const):
    """
    Create a (len(values), 2)-shape array representing (x, y) pairs.

    *values* go into the coordinate determined by ``self.nth_coord``.
    The other coordinate is filled with the constant *const*.

    Example::

        >>> self.nth_coord = 0
        >>> self._to_xy([1, 2, 3], const=0)
        array([[1, 0],
               [2, 0],
               [3, 0]])
    """

Comment (anntzer) [reply lib/mpl_toolkits/axisartist/axislines.py:111]: Yes, that looks like a good description.

File: lib/mpl_toolkits/axisartist/axislines.py [hunk 2]
```diff
@@ -225,8 +237,7 @@ def get_tick_iterators(self, axes):
 
             def _f(locs, labels):
                 for x, l in zip(locs, labels):
```
Comment (timhoffm) [inline lib/mpl_toolkits/axisartist/axislines.py:239]: Optional, but
```suggestion
                for loc, label in zip(locs, labels):
```
would be a bit clearer. in `self._to_xy(x, const=self._pos)` *x* can be a y-coordinate depending on `self._pos`, so better use `self._to_xy(loc, const=self._pos)`.

File: lib/mpl_toolkits/axisartist/axislines.py [hunk 3]
```diff
@@ -298,8 +303,7 @@ def get_tick_iterators(self, axes):
...
```
Comment (timhoffm) [inline lib/mpl_toolkits/axisartist/axislines.py:305]: Optional as above:
```suggestion
                for loc, label in zip(locs, labels):
```

**Review by timhoffm — APPROVED** [review]

**Review by QuLogic — CHANGES_REQUESTED** [review]
This appears to make a different `Path` for `loc='right', nth_coord=0` and `loc='top', nth_coord=1`.

File: lib/mpl_toolkits/axisartist/axislines.py (hunk 2 above)
Comment (anntzer) [reply lib/mpl_toolkits/axisartist/axislines.py:239]: sure

File: lib/mpl_toolkits/axisartist/axislines.py [hunk 4]
```diff
@@ -108,41 +108,53 @@ def update_lim(self, axes):
...
+                array([[1, 0],
+                       [2, 0],
+                       [3, 0]])
+            """
+            if self.nth_coord == 0:
+                return np.stack(np.broadcast_arrays(values, const), axis=-1)
+            elif self.nth_coord == 1:
+                return np.stack(np.broadcast_arrays(const, values), axis=-1)
+            else:
+                raise ValueError("Unxpected nth_coord")
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/mpl_toolkits/axisartist/axislines.py:131]: yes

**Comment by anntzer** [comment]
> This appears to make a different Path for loc='right', nth_coord=0 and loc='top', nth_coord=1.

True, but does a "right x-axis" or a "top y-axis" even make sense?  (to be clear, previously a right x-axis would be a 0-1 x-axis at y=0 and now it's at y=1)  If anything we should just make these error out (which I can do as part of this PR or of a followup).

**Comment by anntzer** [comment]
@QuLogic thoughts about my reply above?

**Comment by anntzer** [comment]
Re-ping @QuLogic per my reply to https://github.com/matplotlib/matplotlib/pull/22314#pullrequestreview-863174000.

File: lib/mpl_toolkits/axisartist/axislines.py [hunk 5]
```diff
@@ -95,38 +95,51 @@ def update_lim(self, axes):
...
-            if nth_coord is None:
-                if loc in ["left", "right"]:
-                    nth_coord = 1
-                else:  # "bottom", "top"
-                    nth_coord = 0
-
-            self.nth_coord = nth_coord
-
+            self._pos = _api.check_getitem(
+                {"bottom": 0, "top": 1, "left": 0, "right": 1}, loc=loc)
```
Comment (tacaswell) [inline lib/mpl_toolkits/axisartist/axislines.py:136]:
```suggestion
            self._pos = {"bottom": 0, "top": 1, "left": 0, "right": 1}[loc]
```

We check 2 lines above that the string is in the list of 4 acceptable inputs so this should always work.

Comment (anntzer) [reply lib/mpl_toolkits/axisartist/axislines.py]: Indeed, fixed.

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
> True, but does a "right x-axis" or a "top y-axis" even make sense? (to be clear, previously a right x-axis would be a 0-1 x-axis at y=0 and now it's at y=1) If anything we should just make these error out (which I can do as part of this PR or of a followup).

Yes, I'm thinking we should probably start erroring for weird inconsistent input, but that can go in a separate PR.
